File: app/src/main/java/com/ymovie/app/ui/home/HomeViewModel.java
package com.ymovie.app.ui.home;

import com.ymovie.app.data.MovieRepository;
import com.ymovie.app.data.ResponseCallback;
import com.ymovie.app.data.model.movie.MovieList;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

public class HomeViewModel extends ViewModel {
    private MutableLiveData<MovieList> movieLiveData;
    private final MovieRepository movieRepository;

    public HomeViewModel(MovieRepository movieRepository) {
        this.movieRepository = movieRepository;
    }

    public LiveData<MovieList> getMovieLiveData() {
        movieLiveData = new MutableLiveData<>();

        return movieLiveData;
    }

    public void fetchTopRatedMovies(String language, int page, String region) {
        movieRepository.fetchTopRatedMovies(language, page, region, new ResponseCallback<MovieList>() {
            @Override
            public void onSuccess(MovieList data) {
                movieLiveData.setValue(data);
            }

            @Override
            public void onFailure(Throwable t) {
                movieLiveData.setValue(null);
            }
        });
    }
}

class HomeViewModelFactory implements ViewModelProvider.Factory {
    private final MovieRepository movieRepository;

    public HomeViewModelFactory(MovieRepository movieRepository) {
        this.movieRepository = movieRepository;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(HomeViewModel.class)) {
            return (T) new HomeViewModel(movieRepository);
        }

        throw new IllegalArgumentException("Unknown ViewModel class");
    }
}

File: app/src/main/java/com/ymovie/app/ui/home/HomeViewModel.kt
package com.ymovie.app.ui.home

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.ymovie.app.data.MovieRepository
import com.ymovie.app.data.model.HomeReqParam
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn

class HomeViewModel(private val movieRepository: MovieRepository) : ViewModel() {
    private var homeReqParam = MutableStateFlow(HomeReqParam())

    val nowPlayingMoviesUiState: StateFlow<HomeUiState> = homeReqParam.flatMapLatest { param ->
        movieRepository.fetchNowPlayingMovies(param.language, param.page, param.region)
            .catch { e ->
                HomeUiState.Failure(Exception(e))
            }
            .map { data ->
                HomeUiState.Success(data)
            }
    }.stateIn(
        scope = viewModelScope,
        started = SharingStarted.WhileSubscribed(5000),
        initialValue = HomeUiState.Loading
    )

    val popularMoviesUiState: StateFlow<HomeUiState> = homeReqParam.flatMapLatest { param ->
        movieRepository.fetchPopularMovies(param.language, param.page, param.region)
            .catch { e ->
                HomeUiState.Failure(Exception(e))
            }
            .map { data ->
                HomeUiState.Success(data)
            }
    }.stateIn(
        scope = viewModelScope,
        started = SharingStarted.WhileSubscribed(5000),
        initialValue = HomeUiState.Loading
    )

    val topRatedMoviesUiState: StateFlow<HomeUiState> = homeReqParam.flatMapLatest { param ->
        movieRepository.fetchTopRatedMovies(param.language, param.page, param.region)
            .catch { e ->
                HomeUiState.Failure(Exception(e))
            }
            .map { data ->
                HomeUiState.Success(data)
            }
    }.stateIn(
        scope = viewModelScope,
        started = SharingStarted.WhileSubscribed(5000),
        initialValue = HomeUiState.Loading
    )

    val upcomingMoviesUiState: StateFlow<HomeUiState> = homeReqParam.flatMapLatest { param ->
        movieRepository.fetchUpcomingMovies(param.language, param.page, param.region)
            .catch { e ->
                HomeUiState.Failure(Exception(e))
            }
            .map { data ->
                HomeUiState.Success(data)
            }
    }.stateIn(
        scope = viewModelScope,
        started = SharingStarted.WhileSubscribed(5000),
        initialValue = HomeUiState.Loading
    )

    fun setHomeReqParam(param: HomeReqParam) {
        homeReqParam.value = param
    }
}

class HomeViewModelFactory(private val movieRepository: MovieRepository) : ViewModelProvider.Factory {
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(HomeViewModel::class.java)) {
            return HomeViewModel(movieRepository) as T
        }

        throw IllegalArgumentException("Unknown ViewModel class")
    }
}